### ISSUE TITLE
Issues adding AWS ECR repository

### ISSUE THREAD

**Issue by awsiv** [spinnaker/spinnaker#4801]
### Issue Summary:
Issues adding AWS ECR repository using halyard

### Cloud Provider(s):
AWS

### Environment:
Spinnaker on EKS using halyard

### Feature Area (if this issue is UI/UX related, please tag `@spinnaker/ui-ux-team`):
`halyard` `component/clouddriver` 

### Description:
Unable to add ECR repo using passwordCommand. I have confirmed that the role I am using i can `list-images` in the ECR repo.

halyard config:
```
    dockerRegistry:
      enabled: true
      accounts:
      - name: test-ecr
        requiredGroupMembership: []
        providerVersion: V1
        permissions: {}
        address: https://11111111.dkr.ecr.<region>.amazonaws.com
        username: AWS
        passwordCommand: aws --region <region> ecr get-authorization-token --registry-ids 11111111 --output text --query 'authorizationData[].authorizationToken' | base64 -d | sed 's/^AWS://'
        cacheIntervalSeconds: 30
        clientTimeoutMillis: 60000
        cacheThreads: 1
        paginateSize: 100
        sortTagsByDate: true
        trackDigests: false
        insecureRegistry: false
        repositories: []
      primaryAccount: test-ecr
```

```
2019-08-27 11:59:38.272  WARN 1 — [cutionAction-36] c.n.s.c.d.r.a.v.c.DockerRegistryClient   : Error encountered during catalog of null
retrofit.RetrofitError: 403 Forbidden
at retrofit.RetrofitError.httpError(RetrofitError.java:40) ~[retrofit-1.9.0.jar:na]
at retrofit.RestAdapter$RestHandler.invokeRequest(RestAdapter.java:388) ~[retrofit-1.9.0.jar:na]
at retrofit.RestAdapter$RestHandler.invoke(RestAdapter.java:240) ~[retrofit-1.9.0.jar:na]
at com.sun.proxy.$Proxy142.getCatalog(Unknown Source) ~[na:na]
at com.netflix.spinnaker.clouddriver.docker.registry.api.v2.client.DockerRegistryClient$DockerRegistryService$getCatalog.call(Unknown Source) ~[na:na]
at com.netflix.spinnaker.clouddriver.docker.registry.api.v2.client.DockerRegistryClient$_getCatalog_closure9.doCall(DockerRegistryClient.groovy:333) ~[clouddriver-docker.jar:na]
at com.netflix.spinnaker.clouddriver.docker.registry.api.v2.client.DockerRegistryClient$_getCatalog_closure9.doCall(DockerRegistryClient.groovy) ~[clouddriver-docker.jar:na]
at sun.reflect.GeneratedMethodAccessor1317.invoke(Unknown Source) ~[na:na]
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_212]
at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_212]
at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101) [groovy-2.5.7.jar:2.5.7]
at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323) [groovy-2.5.7.jar:2.5.7]
at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:263) ~[groovy-2.5.7.jar:2.5.7]
at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041) ~[groovy-2.5.7.jar:2.5.7]
at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:37) ~[groovy-2.5.7.jar:2.5.7]
at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:119) ~[groovy-2.5.7.jar:2.5.7]
at com.netflix.spinnaker.clouddriver.docker.registry.api.v2.client.DockerRegistryClient.request(DockerRegistryClient.groovy:422) ~[clouddriver-docker.jar:na]
at com.netflix.spinnaker.clouddriver.docker.registry.api.v2.client.DockerRegistryClient$request$0.callCurrent(Unknown Source) ~[na:na]
at com.netflix.spinnaker.clouddriver.docker.registry.api.v2.client.DockerRegistryClient.getCatalog(DockerRegistryClient.groovy:331) ~[clouddriver-docker.jar:na]
at com.netflix.spinnaker.clouddriver.docker.registry.api.v2.client.DockerRegistryClient.getCatalog(DockerRegistryClient.groovy) ~[clouddriver-docker.jar:na]
at com.netflix.spinnaker.clouddriver.docker.registry.api.v2.client.DockerRegistryClient$getCatalog.call(Unknown Source) ~[na:na]
at com.netflix.spinnaker.clouddriver.docker.registry.security.DockerRegistryCredentials.getRepositories(DockerRegistryCredentials.groovy:65) ~[clouddriver-docker.jar:na]
at sun.reflect.GeneratedMethodAccessor1316.invoke(Unknown Source) ~[na:na]
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_212]
at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_212]
at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:101) [groovy-2.5.7.jar:2.5.7]
at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323) [groovy-2.5.7.jar:2.5.7]
at org.codehaus.groovy.runtime.metaclass.MethodMetaProperty$GetBeanMethodMetaProperty.getProperty(MethodMetaProperty.java:76) [groovy-2.5.7.jar:2.5.7]
at org.codehaus.groovy.runtime.callsite.GetEffectivePogoPropertySite.callGroovyObjectGetProperty(GetEffectivePogoPropertySite.java:68) [groovy-2.5.7.jar:2.5.7]
at com.netflix.spinnaker.clouddriver.docker.registry.provider.agent.DockerRegistryImageCachingAgent.loadTags(DockerRegistryImageCachingAgent.groovy:90) [clouddriver-docker.jar:na]
at com.netflix.spinnaker.clouddriver.docker.registry.provider.agent.DockerRegistryImageCachingAgent.loadData(DockerRegistryImageCachingAgent.groovy:74) [clouddriver-docker.jar:na]
at com.netflix.spinnaker.cats.agent.CachingAgent$CacheExecution.executeAgentWithoutStore(CachingAgent.java:85) [cats-core.jar:na]
at com.netflix.spinnaker.cats.agent.CachingAgent$CacheExecution.executeAgent(CachingAgent.java:75) [cats-core.jar:na]
at com.netflix.spinnaker.cats.redis.cluster.ClusteredAgentScheduler$AgentExecutionAction.execute(ClusteredAgentScheduler.java:335) [cats-redis.jar:na]
at com.netflix.spinnaker.cats.redis.cluster.ClusteredAgentScheduler$AgentJob.run(ClusteredAgentScheduler.java:305) [cats-redis.jar:na]
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_212]
at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_212]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_212]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_212]
at java.lang.Thread.run(Thread.java:748) [na:1.8.0_212]
2019-08-27 11:59:38.272  INFO 1 — [cutionAction-36] .d.r.p.a.DockerRegistryImageCachingAgent : Describing items in test-ecr/DockerRegistryImageCachingAgent[1/1]
2019-08-27 11:59:38.272  INFO 1 — [cutionAction-36] .d.r.p.a.DockerRegistryImageCachingAgent : Caching 0 tagged images in test-ecr/DockerRegistryImageCachingAgent[1/1]
2019-08-27 11:59:38.272  INFO 1 — [cutionAction-36] .d.r.p.a.DockerRegistryImageCachingAgent : Caching 0 image ids in test-ecr/DockerRegistryImageCachingAgent[1/1]
```
### Steps to Reproduce:
Add and enable ECR repo. Make sure the used role can list images in the repo.
```
    dockerRegistry:
      enabled: true
      accounts:
      - name: test-ecr
        requiredGroupMembership: []
        providerVersion: V1
        permissions: {}
        address: https://11111111.dkr.ecr.<region>.amazonaws.com
        username: AWS
        passwordCommand: aws --region <region> ecr get-authorization-token --registry-ids 11111111 --output text --query 'authorizationData[].authorizationToken' | base64 -d | sed 's/^AWS://'
        cacheIntervalSeconds: 30
        clientTimeoutMillis: 60000
        cacheThreads: 1
        paginateSize: 100
        sortTagsByDate: true
        trackDigests: false
        insecureRegistry: false
        repositories: []
      primaryAccount: test-ecr
```

`hal deploy apply`
then check for clouddriver pod logs

Additionally, try adding repos as:
```
repositories:
- test/repo
```

halyard will complain:
```
Problems in default.provider.dockerRegistry.test-ecr:
! ERROR Unable to fetch tags from the docker repository:
  test/repo, 403 Forbidden
? Can the provided user access this repository?

```
### Additional Details:
Version:
```
 - 1.15.3 (ExtremelyWickedShockinglyEvilAndVile):
   Changelog: https://gist.github.com/spinnaker-release/bed366b82e09498dbb536c098ac11f14
   Published: Mon Aug 26 23:09:21 CEST 2019
   (Requires Halyard >= 1.17)

```
---

**Comment by awsiv** [comment]
Have also verified that I am not using wrong region or wrong account:

https://docs.aws.amazon.com/AmazonECR/latest/userguide/common-errors-docker.html#error-403

**Comment by awsiv** [comment]
`wget --header="Authorization: Basic $TOKEN"  https://<..>.dkr.ecr.<region>.amazonaws.com/v2/_catalog`

Works inside the clouddriver pod. Not sure how spinnaker is querying for this endpoint?

**Comment by Sluggerman** [comment]
This is how we got it working via command line:

>      $HAL_COMMAND config provider docker-registry enable
        $HAL_COMMAND config provider docker-registry account add aws-ecr \
        --address {{ .Values.awsaccount }}.dkr.ecr.{{ .Values.awsregion }}.amazonaws.com --username AWS \
        --password-command "aws ecr get-authorization-token --region {{ .Values.awsregion }} --output text --query 'authorizationData[].authorizationToken' | base64 -d | sed 's/^AWS://'"

**Comment by awsiv** [comment]
thanks @Sluggerman . I did the same thing actually.

However, I think the problem lies in cross account ECR queries to `v2/_catalog`. It works perfectly fine in the ECR repo is in the same account

I am running spinnaker in Account A and trying to query ECR in account B. 
- Followed this guide to setup cross-account access: https://aws.amazon.com/premiumsupport/knowledge-center/secondary-account-access-ecr/ 
- It *works* if I add the repositories to clouddriver config. I guess then spinnaker only has to do `v2/<repo>/list` instead of `v2/_catalog`
- *Fails* if no repo list is provided

Has anyone tried this setup before? Looks like ECR doesn't allow cross-account `v2/_catalog` :(

**Comment by joaquinclearmetal** [comment]
I have successfully setup cross account, and had to modify ECR repo with this snippet below.  The `XXXXXXXXXXXX` is the Account A.

```json
{
  "Version": "2008-10-17",
  "Statement": [
    {
      "Sid": "k8s-aws-permissions",
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::XXXXXXXXXXXX:root"
      },
      "Action": [
        "ecr:BatchCheckLayerAvailability",
        "ecr:BatchGetImage",
        "ecr:DescribeImages",
        "ecr:DescribeRepositories",
        "ecr:GetDownloadUrlForLayer",
        "ecr:GetLifecyclePolicy",
        "ecr:GetLifecyclePolicyPreview",
        "ecr:ListImages"
      ]
    }
  ]
}
```

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by joaquinclearmetal** [comment]
nope

**Comment by joaquinclearmetal** [comment]
How you stop the spinnakerbot close and ignore issue?

**Comment by joaquinclearmetal** [comment]
@awsiv In our environment, we had a `prod` and `test` aws accounts, where ECR lives in `prod` and Spinnaker lives in `test`.  To get off the ground, we had to edit an ECR repository's permissions and under the field of `AWS account IDs` add the `test` aws account.  Then everything on the test account can access the ECR repository.  This is not super secure or elegant, but it worked.

The other problem I had (never solved) was that the password-command would never work, because the cluster didn't have a role that allowed access (and not sure how to add a cross-account role).  So I used an AWS profile key id/secret from the `prod` account.  I didn't know how to get this profiled used with clouddriver, so I manually updated the secret auth token.  Likely, I would have implemented the side car container, would have the AWS profile in the container, e.g. env vars maybe.
